<a href="https://colab.research.google.com/github/chernandezrojas1991/project-precipitation-by-day-chile/blob/main/project_2_ml_precipitation_by_day.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
filename = "/content/drive/MyDrive/Coding Dojo/Machine Learning/Proyecto 2/dataset.csv"
df_dirty = pd.read_csv(filename, header = None)
df_precipitaciones = pd.DataFrame(df_dirty.transpose().values)

<ipython-input-3-d6a76a0d4894>:2: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263

In [ ]:
df_precipitaciones.columns = df_precipitaciones.iloc[0]
df_precipitaciones = df_precipitaciones[1:]
df_precipitaciones.reset_index(drop=True, inplace=True)
df_precipitaciones.head(10)

0 codigo_estacion institucion   fuente                       nombre altura  \
0        01000005         DGA  dga_web                      Visviri   4080   
1        01200002         DGA  dga_web    Villa Industrial (Tacora)   4080   
2        01200003         DGA  dga_web                    Humapalca   3980   
3        01201005         DGA  dga_web  Rio Caracarani En Humapalca   3908   
4        01201010         DGA  dga_web                    Alcerreca   3990   
5        01201003         DGA  dga_web       Rio Lluta En Alcerreca   3550   
6        01001005         DGA  dga_web                      Caquena   4400   
7        01110001         DGA  dga_web                      Puquios   3750   
8        01202009         DGA  dga_web                      Pacollo   4185   
9        01020014         DGA  dga_web                   Cotacotani   4550   

0   latitud  longitud codigo_cuenca             nombre_cuenca  \
0   -17.595  -69.4831            10              Altiplanicas   
1  -17.7719  -69.7244            12                 Rio Lluta   
2   -17.835  -69.7039            12                 Rio Lluta   
3  -17.8428  -69.6994            12                 Rio Lluta   
4  -17.9931  -69.6594            12                 Rio Lluta   
5  -18.0036  -69.6331            12                 Rio Lluta   
6  -18.0542  -69.2017            10              Altiplanicas   
7  -18.1747  -69.7439            11  Quebrada De La Concordia   
8  -18.1769  -69.5092            12                 Rio Lluta   
9  -18.1836  -69.2328            10              Altiplanicas   

0 codigo_sub_cuenca  ... 2020-03-22 2020-03-23 2020-03-24 2020-03-25  \
0               100  ...        0.0        0.6        0.0        4.4   
1               120  ...    -9999.0    -9999.0    -9999.0    -9999.0   
2               120  ...    -9999.0    -9999.0    -9999.0    -9999.0   
3               120  ...        0.2        2.7        3.8        0.0   
4               120  ...    -9999.0    -9999.0    -9999.0    -9999.0   
5               120  ...        0.0        0.0        0.0        0.0   
6               100  ...    -9999.0    -9999.0    -9999.0    -9999.0   
7               111  ...      -9999      -9999      -9999      -9999   
8               120  ...      -9999      -9999      -9999      -9999   
9               102  ...    -9999.0    -9999.0    -9999.0    -9999.0   

0 2020-03-26 2020-03-27 2020-03-28 2020-03-29 2020-03-30 2020-03-31  
0        4.6        0.2        0.8        0.4        0.0        0.0  
1    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0  
2    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0  
3        0.0        0.0        0.1        0.0        0.0        2.6  
4    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0  
5        0.0        0.0        0.0        0.0        0.0        0.5  
6    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0  
7      -9999      -9999      -9999      -9999      -9999      -9999  
8      -9999      -9999      -9999      -9999      -9999      -9999  
9    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0    -9999.0  

[10 rows x 43935 columns]

Limpiar dataframe para que solamente tenga las fechas desde el 1 enero del 2000 hasta fechas del 2020.

In [ ]:
regex = r'^19\d{2}-\d{2}-\d{2}$'
df_filtered = df_precipitaciones.loc[:, ~df_precipitaciones.columns.str.match(regex)]

Realizar transformación de columna respecto a las fechas registradas. Se crea una columna *fecha* y *mediciones*, con tal que la columna *fecha* tenga cada día registrado y *mediciones* tenga los milímetros de precipitaciones caídos ese día.

Nota: Un valor -9999 para *mediciones* significa que ese día no hubo lluvia.

In [ ]:
df_precip_clean = pd.melt(df_filtered, id_vars=['codigo_estacion', 'institucion', 'fuente', 'nombre', 'altura', 'latitud', 'longitud', 'codigo_cuenca', 'nombre_cuenca', 'codigo_sub_cuenca', 'nombre_sub_cuenca', 'inicio_observaciones', 'fin_observaciones', 'cantidad_observaciones', 'inicio_automatica'], var_name='fecha', value_name='mediciones')
df_precip_clean.head(10)

codigo_estacion institucion   fuente                       nombre altura  \
0        01000005         DGA  dga_web                      Visviri   4080   
1        01200002         DGA  dga_web    Villa Industrial (Tacora)   4080   
2        01200003         DGA  dga_web                    Humapalca   3980   
3        01201005         DGA  dga_web  Rio Caracarani En Humapalca   3908   
4        01201010         DGA  dga_web                    Alcerreca   3990   
5        01201003         DGA  dga_web       Rio Lluta En Alcerreca   3550   
6        01001005         DGA  dga_web                      Caquena   4400   
7        01110001         DGA  dga_web                      Puquios   3750   
8        01202009         DGA  dga_web                      Pacollo   4185   
9        01020014         DGA  dga_web                   Cotacotani   4550   

    latitud  longitud codigo_cuenca             nombre_cuenca  \
0   -17.595  -69.4831            10              Altiplanicas   
1  -17.7719  -69.7244            12                 Rio Lluta   
2   -17.835  -69.7039            12                 Rio Lluta   
3  -17.8428  -69.6994            12                 Rio Lluta   
4  -17.9931  -69.6594            12                 Rio Lluta   
5  -18.0036  -69.6331            12                 Rio Lluta   
6  -18.0542  -69.2017            10              Altiplanicas   
7  -18.1747  -69.7439            11  Quebrada De La Concordia   
8  -18.1769  -69.5092            12                 Rio Lluta   
9  -18.1836  -69.2328            10              Altiplanicas   

  codigo_sub_cuenca                      nombre_sub_cuenca  \
0               100  Entre Limite Peru-Bolivia Y Rio Lauca   
1               120                         Rio Lluta Alto   
2               120                         Rio Lluta Alto   
3               120                         Rio Lluta Alto   
4               120                         Rio Lluta Alto   
5               120                         Rio Lluta Alto   
6               100  Entre Limite Peru-Bolivia Y Rio Lauca   
7               111               Quebrada De La Concordia   
8               120                         Rio Lluta Alto   
9               102                              Rio Lauca   

  inicio_observaciones fin_observaciones cantidad_observaciones  \
0           1968-05-01        2020-03-31                  18258   
1           1975-01-01        2019-11-30                  15794   
2           1971-12-01        2019-11-30                  17362   
3           2013-06-01        2020-03-31                   2211   
4           1971-01-01        2019-11-30                  17563   
5           2001-12-01        2020-03-31                   6368   
6           1970-01-01        2019-12-30                  16129   
7           1975-05-01        1981-05-31                   2105   
8           1978-11-01        1999-05-31                   7395   
9           1961-04-01        2018-05-31                  19947   

  inicio_automatica       fecha mediciones  
0        2017-08-02  2000-01-01        0.0  
1                 -  2000-01-01        2.0  
2                 -  2000-01-01    -9999.0  
3        2017-04-04  2000-01-01      -9999  
4                 -  2000-01-01        0.0  
5        2017-09-15  2000-01-01      -9999  
6                 -  2000-01-01        2.0  
7                 -  2000-01-01      -9999  
8                 -  2000-01-01    -9999.0  
9                 -  2000-01-01        8.0

In [ ]:
df_precip_clean['nombre_sub_cuenca'].value_counts()

nombre_sub_cuenca
Rio Loa Alto (Bajo Junta Rio Salado)           192296
Rio Claro                                      133128
Rio Mapocho Bajo                               133128
Rio Serrano                                    118336
Cachapoal Bajo                                 110940
                                                ...  
Costera Quilimari - Petorca                      7396
Costeras Entre Estero Pupio Y Rio Quilimari      7396
Rio Huatulame (Muro Embalse Paloma)              7396
E.  El Teniente Y E. El Almendro (Incl.)         7396
No Disponible                                    7396
Name: count, Length: 227, dtype: int64

# Preguntas a responder

1. Fuente de datos: [Datos de Precipitaciones](https://www.cr2.cl/datos-de-precipitacion/)

2. Breve descripción de los datos

Datos observados de precipitación acumulada diaria desde 879 estaciones de todo Chile, durante el período Enero de 1900-Marzo de 2020. Recolectados de sistemas informáticos de diversas fuentes que se describen a continuación. Compilados por el equipo de Datos y Cómputos del (CR)2 a cargo de Francisca Muñoz, para fines de investigación y docencia. No sustituyen los datos originales provistos por las instituciones responsables de los mismos.

3. ¿Cuál es el objetivo?

El estudio de este conjunto de datos de las precipitaciones acumuladas semanales tiene como objeto ser capaces de predecir los milímetros de precipitaciones caídos en un año determinado según los datos registrados en años anteriores. La cantidad de agua caída, expresado en milímetros, es importante para las zonas norte, centro y sur de Chile, puesto que en cada una de éstas la cantidad de milímetros es importante. Entonces, al predecir los milímetros de agua caída se puede determinar si hay o no déficit (sequía)

4. ¿Este es un problema de clasificación o regresión?

Es un problema de regresión. Se predecirá valores continuos (numéricos).

5. ¿Cuántas características hay?

17 características.

6. ¿Cuántas filas de datos hay?

Hay 6.501.084 observaciones registradas en cuanto a precipitaciones entre el 1 Enero del 2000 al 31 de diciembre de 2020.

In [ ]:
df_precip_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6501084 entries, 0 to 6501083
Data columns (total 17 columns):
 #   Column                  Dtype 
---  ------                  ----- 
 0   codigo_estacion         object
 1   institucion             object
 2   fuente                  object
 3   nombre                  object
 4   altura                  object
 5   latitud                 object
 6   longitud                object
 7   codigo_cuenca           object
 8   nombre_cuenca           object
 9   codigo_sub_cuenca       object
 10  nombre_sub_cuenca       object
 11  inicio_observaciones    object
 12  fin_observaciones       object
 13  cantidad_observaciones  object
 14  inicio_automatica       object
 15  fecha                   object
 16  mediciones              object
dtypes: object(17)
memory usage: 843.2+ MB


7. ¿Qué retos, si es que hay alguno, prevén en la limpieza, exploración o modelización con este conjunto de datos?

La agrupación de los datos en cuanto a fechas. Si bien, estos se pueden agrupar por año o por los meses de invierno por año, que es donde llueve más. La limpieza de los datos implicó realizar traspuesta, dado que los datos están guardados de tal manera que al cargar en Colab, lo que debía ser una observación (o fila) se alojaba en forma de columna.